workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java

```diff
@@ -0,0 +1,131 @@
+package bio.terra.tanagra.workflow;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Flatten;
+import org.apache.beam.sdk.transforms.KvSwap;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.join.CoGbkResult;
+import org.apache.beam.sdk.transforms.join.CoGroupByKey;
+import org.apache.beam.sdk.transforms.join.KeyedPCollectionTuple;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.TupleTag;
+
+/** Utilities for working with graphs in Apache Beam. */
+public final class GraphUtils {
+  private GraphUtils() {}
+
+  /**
+   * Computes the transitive closure of a directed acyclic graph defined by {@code edges}.
+   *
+   * <p>The returned PCollection has at least one KV for every pair of vertices [X, Y] where X can
+   * reach Y by traversing the input edges. e.g. If the edges are [{A, B}, {B, C}], then the output
+   * will be [{A, B}, {A, C}, {B, C}].
+   *
+   * @param edges A directed acyclic graph where every KV defines a directed edge with a {Head,
+   *     Tail} vertex pair.
+   * @param maxPathLength The maximum path length to try to find in the graph. If there is a path in
+   *     the graph longer than this, not all paths will be found. This is used to bound execution.
+   * @param <T> The type of the vertex in the graph.
+   * @return A PCollection of all reachable vertex pairs, where there may be duplicate pairs.
+   */
+  public static <T> PCollection<KV<T, T>> transitiveClosure(
+      PCollection<KV<T, T>> edges, int maxPathLength) {
+    // allPaths[n] is the set of vertices {V0, V1}, where V0 -> V1 is reachable in N or less edges.
+    // That is, there is a path with N or less edges from V0 to V1.
+    Map<Integer, PCollection<KV<T, T>>> allPaths = new HashMap<>();
+    // exactPaths[n] is the set of vertices where there exists a path from V0 -> V1 in exactly N
+    // edges.
+    Map<Integer, PCollection<KV<T, T>>> exactPaths = new HashMap<>();
+
+    // The input edges are both allPaths[1] and exactPaths[1].
+    allPaths.put(1, edges);
+    exactPaths.put(1, edges);
+
+    // Iteratively build up allPaths in O(log_2(maxPathLength)) joins. We want to avoid doing many
```

Comment:
Nice algorithm. Is there a name for this or was this something you just fit to purpose for this? May be worth linking a wikipedia article or something for visual learners, if there is any such documentation

Reply:
Found internally in a google doc initially, but added a link to a paper reference. No great graphics unfortunately.